src/rate_limiter.rs

```diff
@@ -136,16 +145,103 @@ impl RateLimiter {
         self.update();
 
         let used = self.used.get(client_ip).map_or(0, |used| *used);
-        let allowance = CLIENT_CAPACITY - used;
+        let new_balance = used + bytes_to_add;
 
-        if bytes_to_add > allowance {
+        if new_balance > self.client_allowance(client_ip) {
             return Err(RoutingError::ExceedsRateLimit(*hash));
         }
 
-        let _ = self.used.insert(*client_ip, used + bytes_to_add);
+        if overcharged {
+            match self.overcharged.entry(*msg_id) {
+                // If there's no overcharge recorded, record it.
+                Entry::Vacant(entry) => {
+                    let _ = entry.insert(bytes_to_add);
+                }
+                // Disallow a 2nd GET with the same message ID.
+                Entry::Occupied(_) => {
+                    return Err(RoutingError::InvalidMessage);
+                }
+            }
+        }
+
+        let _ = self.used.insert(*client_ip, new_balance);
         Ok(bytes_to_add)
     }
 
+    /// Compute the usage limit for any single client at the current point in time.
+    fn client_allowance(&self, client_ip: &IpAddr) -> u64 {
+        let num_clients = if self.used.contains_key(client_ip) {
+            self.used.len()
+        } else {
+            self.used.len() + 1
+        };
+        cmp::max(MIN_CLIENT_CAPACITY, SOFT_CAPACITY / num_clients as u64)
+    }
+
+    /// Update a client's balance to compensate for initial over-counting.
+    ///
+    /// When a request is made, clients are charged at the maximum size of the data
+    /// being requested. This method compensates the client for the over-counting by
+    /// crediting them the difference between the maximum and the actual size of the response.
+    pub fn deduct_response(
+        &mut self,
+        client_ip: &IpAddr,
+        msg_id: &MessageId,
+        part_count: u32,
+        part_index: u32,
+        payload: &[u8],
+    ) -> Option<u64> {
+        use Response::*;
+
+        // Check that the response isn't a single-part response that we never overcharged for.
+        // This prevents a malicious client from discounting their GET by preceding
+        // it with a PUT with the same `msg_id`.
```

Comment:
any other small charged request, say SetMDataUserPermissions or InsAuthKey, can achieve the same effect, right?